File: src/main/java/com/mypalantir/reasoning/function/builtin/DetectLateUpload.java
package com.mypalantir.reasoning.function.builtin;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.*;

/**
 * 检测是否存在门架流水接收时间晚于拆分时间的情况。
 * 输入：gantry_transactions (list), split_time (date string)
 */
public class DetectLateUpload extends AbstractBuiltinFunction {

    private static final DateTimeFormatter[] FORMATTERS = {
        DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"),
        DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss"),
        DateTimeFormatter.ISO_LOCAL_DATE_TIME
    };

    @Override
    public String getName() { return "detect_late_upload"; }

    @Override
    public Object execute(List<Object> args) {
        List<Map<String, Object>> gantryTxs = asList(args.get(0));
        Object splitTimeArg = args.get(1);

        LocalDateTime splitTime = parseDateTime(splitTimeArg);
        if (splitTime == null) return false;

        for (Map<String, Object> tx : gantryTxs) {
            LocalDateTime receiveTime = parseDateTime(tx.get("receive_time"));
            if (receiveTime != null && receiveTime.isAfter(splitTime)) {
                return true; // 接收时间晚于拆分时间
            }
        }
        return false;
    }

    private LocalDateTime parseDateTime(Object value) {
        if (value == null) return null;
        if (value instanceof LocalDateTime) return (LocalDateTime) value;
        String str = value.toString();
        for (DateTimeFormatter fmt : FORMATTERS) {
            try { return LocalDateTime.parse(str, fmt); } catch (DateTimeParseException ignored) {}
        }
        return null;
    }
}

File: ontology/functions/script/obu_split_analysis.ontology/detect_late_upload.js
// 兼容 Java List：Array.isArray() 对 Java List 为 false
function run(gantry_transactions, split_time) {
  if (gantry_transactions == null || typeof gantry_transactions.length !== 'number') return false;
  if (split_time == null) return false;

  function toMillis(x) {
    if (x == null) return NaN;
    if (typeof x === 'number') return x;
    if (typeof x === 'string') return new Date(x).getTime();
    if (x.getTime) return x.getTime(); // JS Date / java.util.Date / Timestamp 等
    // 兼容 java.time.LocalDateTime 等：优先用 toString() 后再解析
    try {
      var s = String(x);
      var ms = new Date(s).getTime();
      if (!isNaN(ms)) return ms;
    } catch (e) {}
    return Number(x);
  }

  var st = toMillis(split_time);
  if (isNaN(st)) return false;

  for (var i = 0; i < gantry_transactions.length; i++) {
    var rt = gantry_transactions[i].receive_time;
    if (rt == null) continue;
    var rtMs = toMillis(rt);
    if (isNaN(rtMs)) continue;
    if (rtMs > st) return true;
  }
  return false;
}
